tools/clang/lib/SPIRV/EmitSPIRVAction.cpp

```diff
@@ -58,14 +99,34 @@ class SPIRVEmitter : public ASTConsumer {
     const uint32_t funcType = translateFunctionType(decl);
     const uint32_t retType = translateType(decl->getReturnType());
 
-    theBuilder.beginFunction(funcType, retType);
+    const uint32_t funcId = theBuilder.beginFunction(funcType, retType);
     // TODO: handle function parameters
     // TODO: handle function body
     const uint32_t entryLabel = theBuilder.bbCreate();
     theBuilder.bbReturn(entryLabel);
     theBuilder.endFunction();
-  }
 
+    // Add an entry point to the module if necessary
+    const std::string hlslEntryFn =
+        theCompilerInstance.getCodeGenOpts().HLSLEntryFunction;
+    if (hlslEntryFn == decl->getNameInfo().getAsString()) {
+      const spv::ExecutionModel em = getSpirvShaderKindFromHlslProfile(
+          theCompilerInstance.getCodeGenOpts().HLSLProfile.c_str());
+      // TODO: Pass correct input/output interfaces to addEntryPoint.
+      theBuilder.addEntryPoint(em, funcId, hlslEntryFn, {});
+
+      // OpExecutionMode declares an execution mode for an entry point.
+      std::vector<uint32_t> execModeInstr;
```

Comment:
Originally I'm more leaning towards to hide all instruction building logic inside the `ModuleBuilder`; the frontend action should not directly see and use `InstBuilder` and it should use methods exported by `ModuleBuilder`, which acts as a higher level wrapping around the `InstBuilder`. That's a better segregation of concerns. But I see the pain here since we can have additional parameters for `OpExecutionMode`, which means we need to have serveral overloads for adding execution mode. So I'm a little bit on the fence now.

We should add a `take()` method in `InstBuilder` to return the instruction under construction to simplify the code here as suggested in #43.